File: KX-Player/src/stores/taxonomy.ts
/** 分类与标签 store（P0-11/12/13）：分类树、标签云、track↔tag/category 映射与打标操作。 */
import { defineStore } from 'pinia'
import { ref, shallowRef, computed } from 'vue'
import { api } from '@/bridge/ipc'
import { useUiStore } from '@/stores/ui'
import type { Category, Tag, TagSuggestion } from '@/contracts/api'

export const useTaxonomyStore = defineStore('taxonomy', () => {
  const categories = shallowRef<Category[]>([])
  const tags = shallowRef<Tag[]>([])
  /** tagId → trackIds（筛选用，懒加载后维护） */
  const tracksByTag = shallowRef<Map<number, Set<string>>>(new Map())
  /** categoryId → trackIds（含子孙） */
  const tracksByCategory = shallowRef<Map<number, Set<string>>>(new Map())
  const loaded = ref(false)

  function flatten(nodes: Category[], out: Category[] = []): Category[] {
    for (const n of nodes) {
      out.push(n)
      flatten(n.children, out)
    }
    return out
  }

  const flatCategories = computed<Category[]>(() => flatten(categories.value))

  function rebuildMappings(): void {
    void api.taxonomyAllTrackTags().then((rows) => {
      const m = new Map<number, Set<string>>()
      for (const { trackId, tagId } of rows) {
        if (!m.has(tagId)) m.set(tagId, new Set())
        m.get(tagId)!.add(trackId)
      }
      tracksByTag.value = m
    })
    void api.taxonomyAllCategoryItems().then((rows) => {
      // 子孙分类合并（tree 模式）
      const direct = new Map<number, Set<string>>()
      for (const { categoryId, trackId } of rows) {
        if (!direct.has(categoryId)) direct.set(categoryId, new Set())
        direct.get(categoryId)!.add(trackId)
      }
      const m = new Map<number, Set<string>>()
      for (const cat of flatten(categories.value)) {
        const ids = new Set<string>(direct.get(cat.id) ?? [])
        for (const child of flatten(cat.children)) {
          for (const t of direct.get(child.id) ?? []) ids.add(t)
        }
        m.set(cat.id, ids)
      }
      tracksByCategory.value = m
    })
  }

  async function refresh(): Promise<void> {
    categories.value = await api.taxonomyListCategories()
    tags.value = await api.taxonomyListTags()
    rebuildMappings()
    loaded.value = true
  }

  const ui = useUiStore()

  /** 某条目已打的标签 id 集（右键菜单勾选态） */
  function tagsOfTrack(trackId: string): number[] {
    const out: number[] = []
    for (const [tagId, ids] of tracksByTag.value) {
      if (ids.has(trackId)) out.push(tagId)
    }
    return out
  }

  async function createCategory(parentId: number | null, name: string): Promise<void> {
    const cat = await api.taxonomyCreateCategory(parentId, name)
    if (cat) {
      ui.toast(`已创建分类「${name}」`, 'success')
      await refresh()
    } else {
      ui.toast('创建分类失败', 'error')
    }
  }

  async function renameCategory(id: number, name: string): Promise<void> {
    if (await api.taxonomyRenameCategory(id, name)) {
      await refresh()
    } else {
      ui.toast('分类重命名失败，名称可能已存在', 'error')
    }
  }

  async function deleteCategory(id: number): Promise<void> {
    if (await api.taxonomyDeleteCategory(id)) {
      // 当前筛选命中的分类（或其子孙）被删除时，清空分类筛选，避免悬空 id 导致结果恒为空
      if (ui.categoryId !== null) {
        const doomed = new Set<number>([id])
        const collect = (nodes: Category[]): void => {
          for (const n of nodes) {
            doomed.add(n.id)
            collect(n.children)
          }
        }
        const findNode = (nodes: Category[]): Category | null => {
          for (const n of nodes) {
            if (n.id === id) return n
            const hit = findNode(n.children)
            if (hit) return hit
          }
          return null
        }
        const target = findNode(categories.value)
        if (target) collect(target.children)
        if (doomed.has(ui.categoryId)) ui.categoryId = null
      }
      ui.toast('已删除分类', 'success')
      await refresh()
    }
  }

  async function assignCategory(categoryId: number, trackIds: string[]): Promise<void> {
    if (await api.taxonomyAssignCategory(categoryId, trackIds)) {
      ui.toast(`已加入分类（${trackIds.length} 项）`, 'success')
      await refresh()
    }
  }

  async function unassignCategory(categoryId: number, trackIds: string[]): Promise<void> {
    if (await api.taxonomyUnassignCategory(categoryId, trackIds)) await refresh()
  }

  async function tagTracks(tagIds: number[], trackIds: string[], mode: 'add' | 'replace' = 'add'): Promise<void> {
    if (await api.taxonomyTagTracks(tagIds, trackIds, mode)) {
      ui.toast(`已打标签（${trackIds.length} 项）`, 'success')
      await refresh()
    }
  }

  async function untagTracks(tagId: number, trackIds: string[]): Promise<void> {
    if (await api.taxonomyUntagTracks(tagId, trackIds)) await refresh()
  }

  async function upsertTag(name: string, color?: string | null): Promise<Tag | null> {
    const t = await api.taxonomyUpsertTag(name, color ?? null)
    if (t) {
      await refresh()
    } else {
      ui.toast('新建标签失败', 'error')
    }
    return t
  }

  async function renameTag(id: number, name: string): Promise<void> {
    if (await api.taxonomyRenameTag(id, name)) {
      ui.toast(`已重命名为「${name.trim()}」`, 'success')
      await refresh()
    } else {
      ui.toast('标签重命名失败，名称可能已存在', 'error')
    }
  }

  async function deleteTag(id: number): Promise<void> {
    if (await api.taxonomyDeleteTag(id)) {
      if (ui.tagIds.has(id)) {
        const next = new Set(ui.tagIds)
        next.delete(id)
        ui.tagIds = next
      }
      await refresh()
    }
  }

  /** 自动打标：生成建议 → 弹确认（AutoTagReview 由调用方渲染）→ 应用 */
  async function suggestTags(trackIds: string[]): Promise<{ trackId: string; suggestions: TagSuggestion[] }[]> {
    return api.taxonomySuggestTags(trackIds)
  }

  async function applySuggested(accepted: { trackId: string; tagName: string }[]): Promise<void> {
    if (accepted.length && (await api.taxonomyApplySuggestedTags(accepted))) {
      ui.toast(`已写入 ${accepted.length} 条标签`, 'success')
      await refresh()
    }
  }

  return {
    categories, tags, tracksByTag, tracksByCategory, flatCategories, loaded,
    refresh, tagsOfTrack, createCategory, renameCategory, deleteCategory,
    assignCategory, unassignCategory, tagTracks, untagTracks,
    upsertTag, renameTag, deleteTag, suggestTags, applySuggested,
  }
})

File: KX-Player/src-tauri/src/commands/taxonomy.rs
//! 命令层：分类与标签（taxonomy）。
//!
//! 从 `commands/library.rs` 拆出来（账本 §8 第 3 档 **S5**）：曲库那一半是扫描/进度/书签/监听，
//! 这一半是 19 条分类标签命令，两者的依赖面完全不同（前者要 `AppHandle`/`AppWindowsState`，
//! 后者只要一次 `spawn_blocking`），挤在一个 400+ 行文件里只是历史顺序，不是同一种职责。
//! 业务实现在 `crate::taxonomy`，这里只做参数整形与错误收口。

use crate::error::IpcResult;

use super::join_task;
// ── 分类与标签（同样全部走阻塞池）──────────────────────────────

macro_rules! blocking {
    ($expr:expr) => {
        tauri::async_runtime::spawn_blocking(move || $expr).await
    };
}

#[tauri::command]
pub async fn taxonomy_list_categories() -> Vec<crate::taxonomy::CategoryDto> {
    blocking!(crate::taxonomy::build_category_tree(crate::taxonomy::list_categories())).unwrap_or_default()
}

#[tauri::command]
pub async fn taxonomy_create_category(parent_id: Option<i64>, name: String) -> IpcResult<crate::taxonomy::CategoryDto> {
    let trimmed = name.trim().to_string();
    let id = join_task("新建分类", blocking!(crate::taxonomy::create_category(parent_id, &name)))?;
    Ok(crate::taxonomy::CategoryDto {
        id,
        parent_id,
        name: trimmed,
        icon: None,
        color: None,
        sort_index: 0,
        track_count: 0,
        children: vec![],
    })
}

#[tauri::command]
pub async fn taxonomy_rename_category(id: i64, name: String) -> IpcResult<()> {
    join_task("分类改名", blocking!(crate::taxonomy::rename_category(id, &name)))
}

#[tauri::command]
pub async fn taxonomy_delete_category(id: i64) -> IpcResult<()> {
    join_task("删除分类", blocking!(crate::taxonomy::delete_category(id)))
}

#[tauri::command]
pub async fn taxonomy_move_category(id: i64, parent_id: Option<i64>, sort_index: i64) -> IpcResult<()> {
    join_task("移动分类", blocking!(crate::taxonomy::move_category(id, parent_id, sort_index)))
}

#[tauri::command]
pub async fn taxonomy_assign_category(category_id: i64, track_ids: Vec<String>) -> IpcResult<i64> {
    join_task("归入分类", blocking!(crate::taxonomy::assign_category(category_id, &track_ids)))
}

#[tauri::command]
pub async fn taxonomy_unassign_category(category_id: i64, track_ids: Vec<String>) -> IpcResult<i64> {
    join_task("移出分类", blocking!(crate::taxonomy::unassign_category(category_id, &track_ids)))
}

#[tauri::command]
pub async fn taxonomy_list_tags() -> Vec<crate::taxonomy::TagDto> {
    blocking!(crate::taxonomy::list_tags()).unwrap_or_default()
}

#[tauri::command]
pub async fn taxonomy_upsert_tag(name: String, color: Option<String>, kind: Option<String>) -> IpcResult<crate::taxonomy::TagDto> {
    let trimmed = name.trim().to_string();
    let kind_final = kind.unwrap_or_else(|| "topic".into());
    let color2 = color.clone();
    let kind2 = kind_final.clone();
    let id = join_task("保存标签", blocking!(crate::taxonomy::upsert_tag(&name, color2.as_deref(), &kind2)))?;
    Ok(crate::taxonomy::TagDto {
        id,
        name: trimmed,
        color,
        kind: kind_final,
        use_count: 0,
        auto_generated: false,
    })
}

#[tauri::command]
pub async fn taxonomy_rename_tag(id: i64, name: String) -> IpcResult<()> {
    join_task("标签改名", blocking!(crate::taxonomy::rename_tag(id, &name)))
}

#[tauri::command]
pub async fn taxonomy_delete_tag(id: i64) -> IpcResult<()> {
    join_task("删除标签", blocking!(crate::taxonomy::delete_tag(id)))
}

#[tauri::command]
pub async fn taxonomy_tag_tracks(tag_ids: Vec<i64>, track_ids: Vec<String>, mode: Option<String>) -> IpcResult<i64> {
    join_task(
        "批量打标",
        blocking!(crate::taxonomy::tag_tracks(&tag_ids, &track_ids, mode.as_deref().unwrap_or("add"))),
    )
}

#[tauri::command]
pub async fn taxonomy_untag_tracks(tag_id: i64, track_ids: Vec<String>) -> IpcResult<i64> {
    join_task("取消打标", blocking!(crate::taxonomy::untag_tracks(tag_id, &track_ids)))
}

#[tauri::command]
pub async fn taxonomy_suggest_tags(track_ids: Vec<String>) -> Vec<(String, Vec<crate::taxonomy::TagSuggestionDto>)> {
    blocking!(crate::taxonomy::suggest_tags(&track_ids)).unwrap_or_default()
}

#[tauri::command]
pub async fn taxonomy_apply_suggested_tags(accepted: Vec<(String, String)>) -> IpcResult<i64> {
    // accepted: (trackId, tagName) — 先建 tag 再打标
    join_task(
        "应用建议标签",
        tauri::async_runtime::spawn_blocking(move || -> IpcResult<i64> {
            let mut tag_ids: Vec<i64> = Vec::new();
            let names: Vec<String> = accepted.iter().map(|(_, n)| n.clone()).collect();
            for name in &names {
                let id = crate::taxonomy::upsert_tag(name, None, "topic")?;
                if !tag_ids.contains(&id) {
                    tag_ids.push(id);
                }
            }
            let mut track_ids: Vec<String> = accepted.iter().map(|(t, _)| t.clone()).collect();
            track_ids.dedup();
            crate::taxonomy::tag_tracks(&tag_ids, &track_ids, "add")
        })
        .await,
    )
}

#[tauri::command]
pub async fn taxonomy_category_track_ids(category_id: i64, include_children: Option<bool>) -> Vec<String> {
    blocking!(crate::taxonomy::category_track_ids(category_id, include_children.unwrap_or(true))).unwrap_or_default()
}

#[tauri::command]
pub async fn taxonomy_track_tags(track_ids: Vec<String>) -> std::collections::HashMap<String, Vec<crate::taxonomy::TagDto>> {
    blocking!(crate::taxonomy::tags_of_tracks(&track_ids)).unwrap_or_default()
}

#[tauri::command]
pub async fn taxonomy_all_track_tags() -> Vec<(String, i64)> {
    blocking!(crate::taxonomy::all_track_tags()).unwrap_or_default()
}

#[tauri::command]
pub async fn taxonomy_all_category_items() -> Vec<(i64, String)> {
    blocking!(crate::taxonomy::all_category_items()).unwrap_or_default()
}
